### PR TITLE
add epochId to sendEmail request

### PR DESCRIPTION
**Pull request by sshmm**
### <samp>🤖 Generated by Copilot at a68f476</samp>

This pull request adds the ability to scroll to a specific epoch in the history page by using the URL parameter `epochId`. It also adds the `epoch_id` parameter to various email functions and templates to display the relevant epoch information to the recipients. It modifies the files `api-lib/postmark.ts`, `api-test/hasura/cron/epochs_emails_stubbed.test.ts`, `api/hasura/cron/epochs.ts`, `src/pages/HistoryPage/CurrentEpochPanel.tsx`, `src/pages/HistoryPage/EpochPanel.tsx`, `src/pages/HistoryPage/HistoryPage.tsx`, `src/pages/HistoryPage/Notes.tsx`, and `src/routes/routes.tsx` to implement these features.

## Why
We want to be able to add a link to the epoch notes in the email notification
if the epoch link is used app.coordinape.com/circles/[circleId]/epochs/[epochId] it will switch to the epoch page and scroll to the epoch then expand the notes

### <samp>🤖 Generated by Copilot at a68f476</samp>

*  Add `epoch_id` parameter to email sending functions and pass it to Postmark API to populate email templates with epoch information ([link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-0ae55db9a83e75892b84b9d54fe9f1d61d500dadab7966eb9b161e8a92bc8954R84),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-0ae55db9a83e75892b84b9d54fe9f1d61d500dadab7966eb9b161e8a92bc8954R90),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-0ae55db9a83e75892b84b9d54fe9f1d61d500dadab7966eb9b161e8a92bc8954L101-R107),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-0ae55db9a83e75892b84b9d54fe9f1d61d500dadab7966eb9b161e8a92bc8954L114-R122),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ac1f75a435bbaedbc21a6db941465f7d22fd4da453de64217474857c2be131a5R345),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ac1f75a435bbaedbc21a6db941465f7d22fd4da453de64217474857c2be131a5R407),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ac1f75a435bbaedbc21a6db941465f7d22fd4da453de64217474857c2be131a5R596),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ac1f75a435bbaedbc21a6db941465f7d22fd4da453de64217474857c2be131a5R852),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ac1f75a435bbaedbc21a6db941465f7d22fd4da453de64217474857c2be131a5R858),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ac1f75a435bbaedbc21a6db941465f7d22fd4da453de64217474857c2be131a5R873),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ac1f75a435bbaedbc21a6db941465f7d22fd4da453de64217474857c2be131a5R884),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ac1f75a435bbaedbc21a6db941465f7d22fd4da453de64217474857c2be131a5R893))
*  Modify mock epoch ids and add `epoch_id` parameter to mock email sending functions and expected results in `api-test/hasura/cron/epochs_emails_stubbed.test.ts` to match actual functions and test email sending logic ([link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e5a13fa0ec57b6e4622a89fdea5b914d5b18a0eb540c60318a3ababd4b31425bR50),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e5a13fa0ec57b6e4622a89fdea5b914d5b18a0eb540c60318a3ababd4b31425bL55-R69),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e5a13fa0ec57b6e4622a89fdea5b914d5b18a0eb540c60318a3ababd4b31425bL128-R137),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e5a13fa0ec57b6e4622a89fdea5b914d5b18a0eb540c60318a3ababd4b31425bL136-R145),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e5a13fa0ec57b6e4622a89fdea5b914d5b18a0eb540c60318a3ababd4b31425bR212),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e5a13fa0ec57b6e4622a89fdea5b914d5b18a0eb540c60318a3ababd4b31425bR230),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e5a13fa0ec57b6e4622a89fdea5b914d5b18a0eb540c60318a3ababd4b31425bR236),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e5a13fa0ec57b6e4622a89fdea5b914d5b18a0eb540c60318a3ababd4b31425bR252),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e5a13fa0ec57b6e4622a89fdea5b914d5b18a0eb540c60318a3ababd4b31425bR258))
*  Add `targetEpoch` prop and `targetEpochRef` variable to `CurrentEpochPanel` and `EpochPanel` components and use `useEffect` hook to scroll the target epoch panel into view if the prop is true ([link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-97df18eacb52adcf725974367a4b1007c27644225338253660b03fa66f0b15a4L1-R1),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-97df18eacb52adcf725974367a4b1007c27644225338253660b03fa66f0b15a4R41),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-97df18eacb52adcf725974367a4b1007c27644225338253660b03fa66f0b15a4L53-R56),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-97df18eacb52adcf725974367a4b1007c27644225338253660b03fa66f0b15a4R69-R74),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e3d27b7ad7e6b16d694fc88c0523663c141c03f322914062c6db5cab9e160096R1-R2),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e3d27b7ad7e6b16d694fc88c0523663c141c03f322914062c6db5cab9e160096R21),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e3d27b7ad7e6b16d694fc88c0523663c141c03f322914062c6db5cab9e160096L25-R31),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-e3d27b7ad7e6b16d694fc88c0523663c141c03f322914062c6db5cab9e160096R43-R48))
*  Use `useParams` hook to get `epochId` parameter from URL and use `useMemo` hook to calculate `targetEpochPage` based on `pastEpochs` data and `epochId` in `HistoryPage` component ([link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-2cf00674fa58cba2a035c641f5f18d43d2e823d79c558fefbe79f01e1043478fR8),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-2cf00674fa58cba2a035c641f5f18d43d2e823d79c558fefbe79f01e1043478fR52-R53),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-2cf00674fa58cba2a035c641f5f18d43d2e823d79c558fefbe79f01e1043478fR93-R102))
*  Use `useEffect` hook to set `page` state to `targetEpochPage` value in `HistoryPage` component to show the correct page that contains the target epoch ([link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-2cf00674fa58cba2a035c641f5f18d43d2e823d79c558fefbe79f01e1043478fR131-R134))
*  Pass `targetEpoch` prop to `CurrentEpochPanel` and `EpochPanel` components in `HistoryPage` component and set it to true if the epoch id matches the `epochId` variable ([link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-2cf00674fa58cba2a035c641f5f18d43d2e823d79c558fefbe79f01e1043478fR375),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-2cf00674fa58cba2a035c641f5f18d43d2e823d79c558fefbe79f01e1043478fR410))
*  Wrap `NotesSection` component with `forwardRef` function and add `targetEpoch` prop and `ref` parameter to receive them from parent component and use them to set the default tab and the ref for the scrolling logic ([link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ff5c26b2ff0ffb35226ecc2bc40296132c8d7af40ea9881945382d154e9e722dL2-R2),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ff5c26b2ff0ffb35226ecc2bc40296132c8d7af40ea9881945382d154e9e722dL14-R29),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ff5c26b2ff0ffb35226ecc2bc40296132c8d7af40ea9881945382d154e9e722dL33-R38),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ff5c26b2ff0ffb35226ecc2bc40296132c8d7af40ea9881945382d154e9e722dL46-R55),[link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-ff5c26b2ff0ffb35226ecc2bc40296132c8d7af40ea9881945382d154e9e722dL135-R144))
*  Add a new route to `LoggedInRoutes` component to enable the user to access the history page with a specific epoch id in the URL ([link](https://github.com/coordinape/coordinape/pull/2353/files?diff=unified&w=0#diff-0005a5feea8701ddb5aef0d57e1a5a03c36dae6cd62fc3ecf35fd15a88bb5c03R87))

### CODE REVIEW

**Comment by crabsinger** [comment]
@sshmm can you provide a URL example of a link that scrolls to the epoch in question? Since we do paging, what happens if the epoch is not on the first page?

**Comment by sshmm** [comment]
> @sshmm can you provide a URL example of a link that scrolls to the epoch in question? Since we do paging, what happens if the epoch is not on the first page?

@crabsinger 
if the epoch in a different page, it will move to that page and scroll to the epoch
I am using targetEpochPage to track the epoch page.

example app.coordinape.com/circles/12/epochs/23 if epochId:23 of circle 12 is in the second page 
`   const epochIndex = pastEpochs.findIndex(epoch => epoch.id === epochId);
      if (epochIndex >= 0) {
        return Math.floor(epochIndex / pageSize);
      }
    }`
will be used to determine the page

File: src/pages/HistoryPage/CurrentEpochPanel.tsx [hunk 1]
```diff
@@ -38,6 +38,7 @@ type Props = {
   isAdmin: boolean;
   css?: CSS;
   children?: React.ReactNode;
+  targetEpoch: boolean;
```
Comment (crabsinger) [inline src/pages/HistoryPage/CurrentEpochPanel.tsx:41]: what does this prop mean? like if it is targetted it should be open? @sshmm

Comment (sshmm) [reply src/pages/HistoryPage/CurrentEpochPanel.tsx:41]: yes, Do you think another name will be more fit here?

Comment (crabsinger) [reply src/pages/HistoryPage/CurrentEpochPanel.tsx:41]: yes, it seems you are trying to have a boolean of whether this panel should be expanded or not. targetEpoch is context that doesn't exist within this component, or have meaning as a boolean to me.

How aobut we name it `expanded: boolean` which indicates whether the component should render as expanded or not. @sshmm

**Comment by crabsinger** [comment]
@sshmm CI is failing, lets get this deployed

**Comment by sshmm** [comment]
> @sshmm CI is failing, lets get this deployed

@sshmm  it's passing now after rebasing

**Review by crabsinger — APPROVED** [review]

**Pull request merged**